Always separate on spaces and unprintable characters

arp242 opened the pull request:
Some code is sending our backend mentions with a zero-width joiner for
whatever reason, for example:

    @example\u200d skldjg

The mention that would would get picked up is `@example\u200d`, instead
of the expected `@example`.

This patch always treats spaces and unprintable characters as
a terminator. I'm not sure if it ever makes sense to use a space inside
a mention?

Alternatively, I could add a callback function, or something?
Maintaining a list of all space/unprintable characters doesn't strike me
as a good idea, and the `unicode.Is*()` functions contain some
performance optimisations for the common (non-multibyte) use cases.

The downside is some performance loss:

master:

    BenchmarkGetTags-4       2000000               950 ns/op

this:

    BenchmarkGetTags-4       1000000              1312 ns/op

arp242 commented:
I thought the `v2` branch was merged, but turns out it's not (yet), so it's a bit easier to change the API. Another way to do this would be to change the API to something like this in v2:

    GetTags(prefix rune, str string, terminator func(rune) bool) (tags []Tag)

And you can then use it like:

    mention.GetTags("@", "Hello @foo", func(r rune) bool {
        return unicode.IsSpace(r) || !unicode.IsPrint(r)
    })

Could add a standard function or two as well:

    mention.GetTags("@", "Hello @foo", mention.Space)

Or for spaces and a list of characters (which is the same behaviour as this PR):

    mention.GetTags("@", "Hello @foo", mention.Space("^", "$"))

Which is implemented as:

    func Space(terminator ...rune) {
		return func(r rune) bool {
			for _, t := range terminator {
				if r == t {
					return true
				}
			}
			return unicode.IsSpace(r) || !unicode.IsPrint(r)
		}
	}

This might also make it easier to implement something like #9.

gernest commented:
@Carpetsmoker thanks for finding this. I Think I will resolve the v2 branch issue today. I will merge it upstream and version with tags instead.

Then I will check this PR.

gernest commented:
hey I merged v2 upstream. Can you please resubmit the PR based on the current master?

arp242 committed:
Always separate on spaces and unprintable characters

Some code is sending our backend mentions with a zero-width joiner for
whatever reason, for example:

    @example\u200d skldjg

The mention that would would get picked up is `@example\u200d`, instead
of the expected `@example`.

This patch always treats spaces and unprintable characters as
a terminator. I'm not sure if it ever makes sense to use a space inside
a mention?

Alternatively, I could add a callback function, or something?
Maintaining a list of all space/unprintable characters doesn't strike me
as a good idea, and the `unicode.Is*()` functions contain some
performance optimisations for the common (non-multibyte) use cases.

The downside is some performance loss:

master:

    BenchmarkGetTags-4       2000000               988 ns/op

this patch:

    BenchmarkGetTags-4       1000000              1259 ns/op

arp242 commented:
Thanks! PR is now against current master.

gernest commented:
cool, I'm checking it right now.
